### PR TITLE
Fix: Checkbox is killed by systemd-oomd when running memory/memory_stress_ng job

### LINKED ISSUE #51: LP1974205: Case memory/memory_stress_ng be terminated after performing for a while
This issue was migrated from https://bugs.launchpad.net/plainbox-provider-checkbox/+bug/1974205

## Summary
| Status 	| Created on 	| Heat 	| Importance 	| Security related 	|
|--------	|------------	|------	|------------	|------------------	|
| Confirmed	| 2022-05-19 15:14:17	| 172	| Critical	| False  	|

## Description
[Summary]
The terminal which runs memory/memory_stress_ng job will be closed (It seems that system be reboot) and the process of checkbox-cli be terminated after executing for a while.

Issue can also be observed on Cinnamon Bay platform. It will cause system reboot after running this case for a while.

[Steps to reproduce]
1. Boot into OS
2. Run "checkbox-cli run com.canonical.certification::memory/memory_stress_ng"

[Expected result]
checkbox-cli won't be terminated.

[Actual result]
checkbox-cli will be terminated.

[Failure rate]
10/10

[Additional information]
CID: 202112-29802
SKU: TRBA-DVT2-C4
Base Image: dell-bto-jammy-jellyfish-tentacool-X07-20220331-4.iso
Product Name: XPS 13 9320
BIOS Version: 0.2.13
kernel-version: 5.15.0-23-generic

CID: 202203-30134
SKU: CB16T-DVT2-C1
Image: canonical-oem-somerville-jammy-amd64-20220504-33+jellyfish-minccino+X11
system-manufacturer: Dell Inc.
system-product-name: Precision 7670
bios-version: 1.3.1
CPU: 12th Gen Intel(R) Core(TM) i5-12600HX (16x)
GPU: 00:02.0 VGA compatible controller [0300]: Intel Corporation Device [8086:4688] (rev 0c)
kernel-version: 5.17.0-9004-oem

## Attachments
[sosreport-V-high-memory-2022-06-28-zjcpqwv.tar.xz](https://api.launchpad.net/devel/bugs/1974205/+attachment/5600159/data)
[checkbox-ng.zip](https://api.launchpad.net/devel/bugs/1974205/+attachment/5600160/data)
[sosreport-V-high-memory-2022-06-28-zjcpqwv.tar.xz](https://api.launchpad.net/devel/bugs/1974205/+attachment/5600162/data)

Tags:
['jammy-test', 'oem-priority', 'originate-from-1974073', 'originate-from-1981168', 'originate-from-1981180', 'originate-from-1982914', 'originate-from-1983006', 'originate-from-1983068', 'originate-from-1989083', 'somerville', 'stella', 'sutton']

### PR DESCRIPTION
**Pull request by LiaoU3**
To solve the issue that the stress test stops during memory stress test. Mainly, I add 1 job before the memory stress test (memory/store_and_change_oomd_config) to store the original oomd config and change it to "kill" then reload the config. I also add another job (memory/restore_oomd_config) to restore the original config back.
Fixes #51

### CODE REVIEW

**Comment by pieqq** [comment]
There is one warning shown by shellcheck:

```
  cat ${config_path} | grep ManagedOOMMemoryPressure= | sudo sed -i 's/kill/auto/g' ${config_path}
      ^------------^ SC2002: Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.

For more information:
  https://www.shellcheck.net/wiki/SC2002 -- Useless cat. Consider 'cmd < file...
```

I'm not sure how bad using `cat` is. @yphus do you think @LiaoU3 should follow the advice from the [ShellCheck wiki](https://www.shellcheck.net/wiki/SC2002), or add a comment to have ShellCheck ignore these lines?

**Comment by LiaoU3** [comment]
I've updated update the latest commit to prevent from SC2002 issue.
By the way, I will upload submission after recovering from COVID-19.

**Review by zongminl — COMMENTED** [review]
Could you please elaborate why there's a need to change the order here?

**Review by zongminl — CHANGES_REQUESTED** [review]
- I'd prefer to have a more specific title for this PR, such like "Fix: Checkbox is killed by systemd-oomd when running memory/memory_stress_ng job"
- Please link this PR to the bug #51 
- Please squash commit [87d5b47](https://github.com/canonical/checkbox/pull/297/commits/87d5b473c39a5f6c6d07205f32d07bd67fa7c4af) and [766b159](https://github.com/canonical/checkbox/pull/297/commits/766b1598aa7c9190b506dc6a67341076b4cb1312) into one or you should correct the commit message
- Please use git features to amend your commits instead of creating new commits
- Please read contributing guide in https://github.com/canonical/checkbox/blob/main/CONTRIBUTING.md before creating a PR

File: providers/certification-client/units/client-cert-desktop-22-04.pxu [hunk 1]
```diff
@@ -151,9 +151,9 @@ nested_part:
     # The following tests are purely automated and/or lenghty stress tests.
     # They have been moved to the end of the test run to improve the testing
     # process.
+    stress-ng-cert-automated
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply providers/certification-client/units/client-cert-desktop-22-04.pxu:154]: I tried the original order, the terminal opened after reboot would still cause the issue #51 . It seems like there might be something different between the terminal opened after reboot and the terminal opened by ourselves.

[the remark this answers is not in the document]
Comment (zongminl) [reply providers/certification-client/units/client-cert-desktop-22-04.pxu:154]: Thank you for the details, terminal console started by pm_test.py will be using root permission and this leads to stress-ng running in a more aggressive mode.
This change looks good then.

File: providers/base/units/memory/jobs.pxu [hunk 2]
```diff
@@ -72,7 +72,7 @@ command:
     cat ${config_path}
     echo
     echo "Changing oomd config to auto..."
-    cat ${config_path} | grep ManagedOOMMemoryPressure= | sudo sed -i 's/kill/auto/g' ${config_path}
+    < ${config_path} | grep ManagedOOMMemoryPressure= | sudo sed -i 's/kill/auto/g' ${config_path}
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply providers/base/units/memory/jobs.pxu]: You are right! Your solution is very clear! I have no idea what I was thinking haha.

File: providers/base/units/memory/jobs.pxu [hunk 3]
```diff
@@ -55,3 +55,27 @@ _description:
  swapping to disk, thus SUTs should have suitably large swap files for the
  amount of RAM they have installed.
 
+plugin:shell
+category_id: com.canonical.plainbox::memory
```
[the remark this answers is not in the document]
Comment (LiaoU3) [reply providers/base/units/memory/jobs.pxu]: Got it! It's good to let it be part of **stress** category.

**Comment by LiaoU3** [comment]
@zongminl I've amended my commits. Please review it when you are available!

**Comment by LiaoU3** [comment]
Submissions
Dell : https://certification.canonical.com/hardware/202212-30998/submission/295274/
HP : https://certification.canonical.com/hardware/202203-30089/submission/295275/
Lenovo : Running

File: providers/base/units/stress/jobs.pxu [hunk 4]
```diff
@@ -680,3 +680,52 @@ command:
 environ: WPA_N_SSID WPA_N_PSK
 category_id: com.canonical.plainbox::stress
 flags: preserve-locale also-after-suspend
+
```
Comment (zongminl) [inline providers/base/units/stress/jobs.pxu:683]: Perhaps I didn't make it clear enough with my previous review, but I wish these jobs (`stress/store_and_change_oomd_config` and `stress/restore_oomd_config`) can be moved to stress/stress-ng.pxu

**Review by zongminl — CHANGES_REQUESTED** [review]

File: providers/base/units/stress/jobs.pxu (hunk 4 above)
Comment (LiaoU3) [reply providers/base/units/stress/jobs.pxu]: Copy that. Done.

**Review by zongminl — APPROVED** [review]
LGTM, thank you.

**Comment by LiaoU3** [comment]
@pieqq Hi! need your review and approve! Thanks!

**Review by pieqq — APPROVED** [review]
LGTM!

Once this gets merged, it will be available in the [Checkbox daily builds](https://launchpad.net/~checkbox-dev/+archive/ubuntu/ppa). It would be good to test it "for real" then.

**Comment by bladernr** [comment]
I was hoping this could help with the issues we're seeing around checkbox-remote per the last mention by @pieqq however, looking at the diff, this wont work on at least Ubuntu Server for 20.04 as that file does not exist.  Running these jobs on a focal machine will cause a failure/error messaging, and so the jobs should be constrained to only run on Jammy or later.

**Pull request merged**